core/src/main/java/org/apache/calcite/rel/rules/JoinToHyperGraphRule.java

```diff
@@ -0,0 +1,192 @@
+              .map(hyperEdge -> adjustNodeBit(hyperEdge, leftNodeCount, leftFieldCount))
+              .collect(Collectors.toList()));
+    } else if (left instanceof HyperGraph) {
+      leftNodeCount = left.getInputs().size();
+      inputs.addAll(left.getInputs());
+      inputs.add(right);
+
+      edges.addAll(((HyperGraph) left).getEdges());
+    } else if (right instanceof HyperGraph) {
+      leftNodeCount = 1;
+      inputs.add(left);
+      inputs.addAll(right.getInputs());
+
+      edges.addAll(
+          ((HyperGraph) right).getEdges().stream()
+              .map(hyperEdge -> adjustNodeBit(hyperEdge, leftNodeCount, leftFieldCount))
+              .collect(Collectors.toList()));
+    } else {
+      leftNodeCount = 1;
+      inputs.add(left);
+      inputs.add(right);
+    }
+
+    HashMap<Integer, Integer> fieldIndexToNodeIndexMap = new HashMap<>();
+    int fieldCount = 0;
+    for (int i = 0; i < inputs.size(); i++) {
+      for (int j = 0; j < inputs.get(i).getRowType().getFieldCount(); j++) {
+        fieldIndexToNodeIndexMap.put(fieldCount++, i);
+      }
+    }
+    // convert current join condition to hyper edge condition
+    for (RexNode joinCond : joinConds) {
+      long leftNodeBits;
+      long rightNodeBits;
+      List<Integer> leftRefs = new ArrayList<>();
+      List<Integer> rightRefs = new ArrayList<>();
+
+      RexVisitorImpl visitor = new RexVisitorImpl<Void>(true) {
+        @Override public Void visitInputRef(RexInputRef inputRef) {
+          Integer nodeIndex = fieldIndexToNodeIndexMap.get(inputRef.getIndex());
+          if (nodeIndex == null) {
+            throw new IllegalArgumentException("RexInputRef refers a dummy field: "
+                + inputRef + ", rowType is: " + origJoin.getRowType());
+          }
+          if (nodeIndex < leftNodeCount) {
+            leftRefs.add(nodeIndex);
+          } else {
+            rightRefs.add(nodeIndex);
+          }
+          return null;
+        }
+      };
+      joinCond.accept(visitor);
+
+      // when cartesian product, make it to complex hyper edge
+      if (leftRefs.isEmpty() || rightRefs.isEmpty()) {
+        leftNodeBits = LongBitmap.newBitmapBetween(0, leftNodeCount);
+        rightNodeBits = LongBitmap.newBitmapBetween(leftNodeCount, inputs.size());
+      } else {
+        leftNodeBits = LongBitmap.newBitmapFromList(leftRefs);
+        rightNodeBits = LongBitmap.newBitmapFromList(rightRefs);
+      }
+      edges.add(
+          new HyperEdge(
+              leftNodeBits,
+              rightNodeBits,
+              origJoin.getJoinType(),
+              joinCond));
+    }
+    result =
+        new HyperGraph(
+            origJoin.getCluster(),
+            origJoin.getTraitSet(),
+            inputs,
+            edges,
+            origJoin.getRowType());
+
+    call.transformTo(result);
+  }
+
+  private static HyperEdge adjustNodeBit(HyperEdge hyperEdge, int nodeOffset, int fieldOffset) {
+    RexShuttle shuttle = new RexShuttle() {
```

Comment:
IIANM this shuttle is not needed, you can reuse `RexUtil#shift(RexNode node, final int offset)` method